src/rpc/websocket_api.cpp

```diff
@@ -96,81 +119,120 @@ void websocket_api_connection::send_notice(
                                                    _max_conversion_depth ) );
 }
 
-std::string websocket_api_connection::on_message(
-   const std::string& message,
-   bool send_message /* = true */ )
+response websocket_api_connection::on_message( const std::string& message )
 {
+   variant var;
    try
    {
-      auto var = fc::json::from_string(message, fc::json::legacy_parser, _max_conversion_depth);
-      const auto& var_obj = var.get_object();
+      var = fc::json::from_string( message, fc::json::legacy_parser, _max_conversion_depth );
+   }
+   catch( const fc::exception& e )
+   {
+      return response( variant(), { -32700, "Invalid JSON message", variant( e, _max_conversion_depth ) }, "2.0" );
+   }
 
-      if( var_obj.contains( "method" ) )
-      {
-         auto call = var.as<fc::rpc::request>(_max_conversion_depth);
-         exception_ptr optexcept;
-         try
-         {
-            try
-            {
+   if( var.is_array() )
+      return response( variant(), { -32600, "Batch requests not supported" }, "2.0" );
+
+   if( !var.is_object() )
+      return response( variant(), { -32600, "Invalid JSON request" }, "2.0" );
+
+   variant_object var_obj = var.get_object();
+
+   if( var_obj.contains( "id" )
+       && !var_obj["id"].is_string() && !var_obj["id"].is_numeric() && !var_obj["id"].is_null() )
+      return response( variant(), { -32600, "Invalid id" }, "2.0" );
+
+   if( var_obj.contains( "method" ) && ( !var_obj["method"].is_string() || var_obj["method"].get_string() == "" ) )
+      return response( variant(), { -32602, "Missing or invalid method" }, "2.0" );
```

Comment:
-32601 | Method not found | The method does not exist / is not available.
-- | -- | --

See https://www.jsonrpc.org/specification.

Reply:
Changed to -32600.
(The spec requires the method to be a string value, so anything else is a violation of the spec. An empty string value is more like an absent method name, because an empty method name doesn't really make sense, even if it would technically work.)